File: leetcode/matrix/221. Maximal Square/js/main.js
// Time O(N * M)
// Space O(1)
const maximalSquare = matrix => {
  if (matrix.length == 0) return 0;

  let n = matrix.length;
  let m = matrix[0].length;

  let max = 0;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (matrix[i][j] == '1') {
        if (i == 0 || j == 0) {
          max = Math.max(max, 1);
          continue;
        }

        matrix[i][j] = Math.min(matrix[i][j - 1], matrix[i - 1][j], matrix[i - 1][j - 1]) + 1;
        max = Math.max(max, matrix[i][j]);
      }
    }
  }

  return max * max;
};

// Time O(n*m^2)
// Space O(1)
const maximalSquare_II = function (matrix) {
  const n = matrix.length;
  const m = n > 0 ? matrix[0].length : 0;
  let ans = 0;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (matrix[i][j] === '0') {
        continue;
      }

      let sqlen = 1;
      let flag = true;

      while (flag && i + sqlen < n && j + sqlen < m) {
        // row
        for (let k = j; k <= sqlen + j; k++) {
          if (matrix[i + sqlen][k] === '0') {
            flag = false;
            break;
          }
        }

        // cell
        for (let k = i; k <= sqlen + i; k++) {
          if (matrix[k][j + sqlen] === '0') {
            flag = false;
            break;
          }
        }

        if (flag) {
          sqlen++;
        }
      }

      if (ans < sqlen) {
        ans = sqlen;
      }
    }
  }

  return ans * ans;
};

File: leetcode/matrix/221. Maximal Square/py/main.py
class Solution:
    # Time O(N*M)
    # Space O(N*M)
    def maximalSquare(self, matrix):
        if len(matrix) == 0:
            return 0

        n, m = len(matrix), len(matrix[0])
        res = 0

        dp = [[0 if matrix[i][j] == '0' else 1 for j in range(0, m)]
              for i in range(0, n)]

        for i in range(0, n):
            for j in range(0, m):
                if dp[i][j] == 1:
                    if i == 0 or j == 0:
                        res = max(res, 1)
                    else:
                        dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j],
                                       dp[i][j - 1]) + 1

                        res = max(res, dp[i][j])

        return res * res

    # Time O(n*m^2)
    # Space O(1)
    def maximalSquare_II(self, matrix):
        n, m = len(matrix), len(matrix[0])
        ans = 0

        for i in range(n):
            for j in range(m):
                if matrix[i][j] == "1":
                    length = 1

                    ans = max(ans, length)

                    while i + length <= n and j + length <= m:
                        row, max_col = i, j + length

                        # Row direction
                        while row < i + length and matrix[row][max_col -
                                                               1] == "1":
                            row += 1

                        if row < i + length:
                            break

                        col, max_row = j, i + length

                        # Col direction
                        while col < j + length and matrix[max_row -
                                                          1][col] == "1":
                            col += 1

                        if col < j + length:
                            break

                        ans = max(ans, length)

                        length += 1

        return ans * ans
